In [66]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, accuracy_score, roc_curve
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE
from sklearn.svm import SVC


## First 3 Models
To get an idea of how well our data is, we will build 3 models on features we deemed important based on the literature attached to these datasets. These features include:
-   U:
-   G:
-   R:
-   I:
-   Z:
-   Redshift:
-   ra:
-   dec:


We will train our model on 70% of the data and then test on 30%. The results of this model will indicate to us how we should proceed.


In [10]:
df = pd.read_csv('/Users/ahmadallaou/Documents/UCB Repos/datasci207_project/data/raw/skyserver_sdss/dr16_galaxySpecPhoto.csv')

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_36793/649618218.py:1: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/Users/ahmadallaou/Documents/UCB Repos/datasci207_project/data/raw/skyserver_sdss/dr16_galaxySpecPhoto.csv')


In [11]:
df.columns

Index(['specobjid', 'dr8objid', 'dr7objid', 'ra', 'dec', 'nvote_tot',
       'nvote_std', 'nvote_mr1', 'nvote_mr2', 'nvote_mon', 'p_el', 'p_cw',
       'p_acw', 'p_edge', 'p_dk', 'p_mg', 'p_cs', 'instrument', 'redshift',
       'sdss_class_string', 'sdss_subclass_string', 'class_noqso',
       'subClass_noqso', 'spectroFlux_u', 'spectroFlux_g', 'spectroFlux_r',
       'spectroFlux_i', 'spectroFlux_z', 'spectroSynFlux_u',
       'spectroSynFlux_g', 'spectroSynFlux_r', 'spectroSynFlux_i',
       'spectroSynFlux_z', 'elodieObject', 'elodieSpType', 'elodieBV',
       'elodieTEff', 'elodieLogG', 'elodieFeH', 'elodieZ', 'sdss_class_number',
       'clean', 'u', 'g', 'r', 'i', 'z', 'nObserve', 'nDetect', 'nEdge',
       'score'],
      dtype='object')

In [12]:
features = ['u','g','r','i','z','redshift','ra','dec']

In [13]:
label = ['sdss_class_string']

In [14]:
df['sdss_class_string'].value_counts()

GALAXY    786076
STAR        7328
QSO         7223
Name: sdss_class_string, dtype: int64

In [15]:
X = df[features]
y = df[label]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

clf = RandomForestClassifier(n_estimators=100, random_state=42)

clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_36793/2050606733.py:17: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  clf.fit(X_train, y_train)


0.9931886972342614
              precision    recall  f1-score   support

      GALAXY       0.99      1.00      1.00    235824
         QSO       0.90      0.33      0.49      2167
        STAR       0.98      0.97      0.97      2198

    accuracy                           0.99    240189
   macro avg       0.96      0.77      0.82    240189
weighted avg       0.99      0.99      0.99    240189



In [ ]:
print(accuracy)
print(class_report)

In [23]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

y_pred_knn = knn.predict(X_test)

accuracy_knn = accuracy_score(y_test, y_pred_knn)
class_report_knn = classification_report(y_test, y_pred_knn)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [24]:
print(accuracy_knn)
print(class_report_knn)

0.9877971097760514
              precision    recall  f1-score   support

      GALAXY       0.99      1.00      0.99    235824
         QSO       0.96      0.09      0.17      2167
        STAR       0.77      0.82      0.79      2198

    accuracy                           0.99    240189
   macro avg       0.90      0.64      0.65    240189
weighted avg       0.99      0.99      0.98    240189



In [26]:
logreg = LogisticRegression(max_iter=10000, random_state=42)
logreg.fit(X_train, y_train)

y_pred_logreg = logreg.predict(X_test)

accuracy_logreg = accuracy_score(y_test, y_pred_logreg)
class_report_logreg = classification_report(y_test, y_pred_logreg)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [27]:
print(accuracy_logreg)
print(class_report_logreg)

0.9886297873757749
              precision    recall  f1-score   support

      GALAXY       0.99      1.00      0.99    235824
         QSO       0.95      0.08      0.15      2167
        STAR       0.78      0.93      0.85      2198

    accuracy                           0.99    240189
   macro avg       0.91      0.67      0.66    240189
weighted avg       0.99      0.99      0.99    240189



# Classifying the Objects
Using the SDSS dataset, we've trained a model to classify an object as either a galaxy, quasar or star. We have opted to use several models to classify these objects including (with their respective scores):
-   RFC
    -   Accuracy: 99.3%
    -   Recall: 
        -   Galaxy: 100%
        -   QSO: 33%
        -   STAR: 97%
    -   Precision:
        -   Galaxy: 99%
        -   QSO: 90%
        -   STAR: 98%
    -   F1 Score:
        -   Galaxy: 100%
        -   QSO: 49%
        -   STAR: 97%
<br></br>
-   KNN
    -   Accuracy: 98.8%
    -   Recall: 
        -   Galaxy: 100%
        -   QSO: 9%
        -   STAR: 82%
    -   Precision:
        -   Galaxy: 99%
        -   QSO: 96%
        -   STAR: 77%
    -   F1 Score:
        -   Galaxy: 99%
        -   QSO: 17%
        -   STAR: 79%
<br></br>
-   Logistic Regression
    -   Accuracy: 98.9%
    -   Recall: 
        -   Galaxy: 100%
        -   QSO: 8%
        -   STAR: 93%
    -   Precision:
        -   Galaxy: 99%
        -   QSO: 95%
        -   STAR: 78%
    -   F1 Score:
        -   Galaxy: 99%
        -   QSO: 15%
        -   STAR: 85%



## Interpreting the Results
The accuracy results are phenomenal, however this can be an indication of overfitting. In fact, our recall scores and f1 scores for the Quasar and star objects seem to be quite low in comparison to the galaxy class. This may be due to the fact that our data is imbalanced so we will account for that with various techniques such as SMOTE or randomly selecting the same number of each class during the training. We will also evaluate if PCA might make our model perform better, however this might not be the case since generally an object can realistically be classified just on the redshift feature, since the distance between each of the objects is quite clear based on the literature. We will apply some transformations and perform some EDA and clean the data in the following notebooks:
-   jsm_eda_00.ipynb
-   jsm_eda_01.ipynb
-   jsm_eda_02.ipynb
-   jsm_eda_03.ipynb

In [11]:
sdss_headers = pd.read_csv('../data/processed/sdss_primary.csv', header=0, index_col=0, nrows=0).columns.tolist()

# specify groups of primary columns
object_identifiers = ['dr8objid']
galaxy_identifiers = ['gz2_filenumber']
object_locations = ['ra', 'dec']
object_class_labels = ['sdss_clean_class_name']
star_class_labels = ['stellar_subclass_MK']
galaxy_class_labels =['gz2_subclass_hubble']
galaxy_binary_labels = [label for label in sdss_headers if 'flag' in label]
object_features = ['redshift_final', 'sum_p', 'u-r_p', 'g-i_p', 'r-z_p']
star_features = ['elodie_color_index', 'elodie_temperature', 'elodie_metallicity']
# consider adding galaxy features:  ratio of petro_radius to petro_brightness

# create a datatype dictionary for object_identifiers
col_dtypes = {col: 'string' for col in object_identifiers + galaxy_identifiers}

# load full consolidated sdss dataset
sdss_primary = pd.read_csv('../data/processed/sdss_primary.csv', header=0, dtype=col_dtypes)

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_1993/822661678.py:19: DtypeWarning: Columns (7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46) have mixed types. Specify dtype option on import or set low_memory=False.
  sdss_primary = pd.read_csv('../data/processed/sdss_primary.csv', header=0, dtype=col_dtypes)


In [12]:
sdss_primary

dr8objid gz2_filenumber          ra        dec  \
0        1237651252584448100         192933  236.355530  54.820911   
1        1237651252584448112           <NA>  236.342000  54.802830   
2        1237651252584251555          32264  235.891464  55.133472   
3        1237651252584317089          32265  236.063721  55.035847   
4        1237655130907213985           <NA>  236.150100  54.775720   
...                      ...            ...         ...        ...   
1089268  1237658613059485785           <NA>  186.049290  50.066218   
1089269  1237658613059551315           <NA>  186.294360  50.072488   
1089270  1237658613593866399           <NA>  176.965730  50.196095   
1089271  1237658613593997533           <NA>  177.500070  50.211641   
1089272  1237658613594063058           <NA>  177.796780  50.375821   

        sdss_clean_class_name stellar_subclass_MK gz2_subclass_hubble  \
0                      galaxy        unclassified                 SBc   
1                      galaxy        unclassified        unclassified   
2                      galaxy        unclassified                 SBc   
3                      galaxy        unclassified                  Sc   
4                      galaxy        unclassified        unclassified   
...                       ...                 ...                 ...   
1089268                quasar        unclassified        unclassified   
1089269                  star        unclassified        unclassified   
1089270                quasar        unclassified        unclassified   
1089271                quasar        unclassified        unclassified   
1089272                quasar        unclassified        unclassified   

        gz1_flag_spiral gz1_flag_elliptical gz1_flag_uncertain  ...  \
0                  True               False              False  ...   
1                 False               False               True  ...   
2                  True               False              False  ...   
3                 False               False               True  ...   
4                 False               False               True  ...   
...                 ...                 ...                ...  ...   
1089268             NaN                 NaN                NaN  ...   
1089269             NaN                 NaN                NaN  ...   
1089270             NaN                 NaN                NaN  ...   
1089271             NaN                 NaN                NaN  ...   
1089272             NaN                 NaN                NaN  ...   

        gz2_flag_arms_many gz2_flag_arms_unsure redshift_final      sum_p  \
0                    False                False       0.080127   85.88830   
1                      NaN                  NaN       0.193371   88.50900   
2                    False                False       0.040399   77.52239   
3                    False                False       0.048782   84.18209   
4                      NaN                  NaN       0.066882   89.05857   
...                    ...                  ...            ...        ...   
1089268                NaN                  NaN       0.936257  109.67462   
1089269                NaN                  NaN      -0.000493   83.65337   
1089270                NaN                  NaN       1.325296  102.50690   
1089271                NaN                  NaN       1.076416  108.06205   
1089272                NaN                  NaN       1.856644  107.44989   

           u-r_p    g-i_p    r-z_p elodie_color_index elodie_temperature  \
0        2.09786  1.09673  0.63014                NaN                NaN   
1        2.39652  1.32288  0.71409                NaN                NaN   
2        2.37586  1.12715  0.67207                NaN                NaN   
3        1.77608  0.65633  0.23047                NaN                NaN   
4        1.44878  0.69195  0.46791                NaN                NaN   
...          ...      ...      ...                ...               

In [27]:
x1 = sdss_primary[['dr8objid','redshift_final','sum_p','u-r_p','g-i_p','r-z_p']]
y1 = sdss_primary[['sdss_clean_class_name']]

In [28]:
# Initialize SMOTE
smote = SMOTE()

# Apply SMOTE
X_resampled, y_resampled = smote.fit_resample(x1, y1)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:964: FutureWarning: Arrays of bytes/strings is being converted to decimal numbers if dtype='numeric'. This behavior is deprecated in 0.24 and will be removed in 1.1 (renaming of 0.26). Please convert your data to numeric values explicitly instead.
  X = check_array(


In [29]:
y_resampled.value_counts()

sdss_clean_class_name
galaxy                   531099
quasar                   531099
star                     531099
dtype: int64

In [35]:
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.3, stratify=y_resampled, random_state=42)

id_train = X_train.drop('dr8objid',axis=1)
id_test = X_test.drop('dr8objid',axis=1)

X_train.drop('dr8objid',inplace=True,axis=1)
X_test.drop('dr8objid',inplace=True,axis=1)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

clf2 = RandomForestClassifier(n_estimators=100, random_state=42)

clf2.fit(X_train, y_train)

y_pred = clf2.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_1993/1653946815.py:15: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  clf.fit(X_train, y_train)


In [39]:
print('Accuracy: ',accuracy)

Accuracy:  0.9997510408167535


In [40]:
print(class_report)

              precision    recall  f1-score   support

      galaxy       1.00      1.00      1.00    159330
      quasar       1.00      1.00      1.00    159330
        star       1.00      1.00      1.00    159330

    accuracy                           1.00    477990
   macro avg       1.00      1.00      1.00    477990
weighted avg       1.00      1.00      1.00    477990



## Extreme Overfitting
By oversampling the data as much as it did, the model ended up learning the exact features of each class in the training set, and it was seeing the same rows in the test set, so while the model performed "perfectly" here, this would not be the case for new data since it seems there was some extreme overfitting in this model. Instead what we might do is limit the data to an equal number of each class (randomly sampled) and then train the model using that. 

In [44]:
under = sdss_primary.groupby('sdss_clean_class_name').apply(lambda x: x.sample(n=50000, replace=False,random_state=42))
under = under.reset_index(drop=True)


In [46]:
x_under = under[['dr8objid','redshift_final','sum_p','u-r_p','g-i_p','r-z_p']]
y_under =  under[['sdss_clean_class_name']]

X_train, X_test, y_train, y_test = train_test_split(x_under, y_under, test_size=0.3, stratify=y_under, random_state=42)

id_train = X_train.drop('dr8objid',axis=1)
id_test = X_test.drop('dr8objid',axis=1)

X_train.drop('dr8objid',inplace=True,axis=1)
X_test.drop('dr8objid',inplace=True,axis=1)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

clf3 = RandomForestClassifier(n_estimators=100, random_state=42)

clf3.fit(X_train, y_train)

y_pred = clf3.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_1993/2060985526.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  clf.fit(X_train, y_train)


In [47]:
print('Accuracy: ',accuracy)
print(class_report)

Accuracy:  0.9994888888888889
              precision    recall  f1-score   support

      galaxy       1.00      1.00      1.00     15000
      quasar       1.00      1.00      1.00     15000
        star       1.00      1.00      1.00     15000

    accuracy                           1.00     45000
   macro avg       1.00      1.00      1.00     45000
weighted avg       1.00      1.00      1.00     45000



## Limited Data Trained
We have now trained the model on the limited data (50,000 randomly sampled rows from each class) and it still seems like our results are perfect. To make ensure that these results are valid, we will operationalize the model on the remaining data points and ensure that it is predicting each class correct. 

**NOTE: A reason why the results might be so well is that generally an astronomer can classify an object simply on the redshift because the variance for redshift between each class is so great. This is because redshift is a measure of distance for the objects, and from the literature it is known that stars are the closest, then galaxies and then quasars, and each class is extremely further away than the next.**

In [50]:
remaining = sdss_primary[~sdss_primary['dr8objid'].isin(under['dr8objid'])]

In [55]:
test_50 = remaining.sample(n=50000,random_state=42)
test_50_x = test_50[['redshift_final','sum_p','u-r_p','g-i_p','r-z_p']]
test_50_label = test_50[['sdss_clean_class_name']]

In [56]:
test_50_x = scaler.transform(test_50_x)
y_50_pred = clf.predict(test_50_x)

In [57]:
print(accuracy_score(test_50_label, y_50_pred))
print(classification_report(test_50_label, y_50_pred))

0.9993
              precision    recall  f1-score   support

      galaxy       1.00      1.00      1.00     25738
      quasar       1.00      1.00      1.00      7427
        star       1.00      1.00      1.00     16835

    accuracy                           1.00     50000
   macro avg       1.00      1.00      1.00     50000
weighted avg       1.00      1.00      1.00     50000



In [58]:
rem_label = remaining[['sdss_clean_class_name']]
rem_pred = clf.predict(scaler.transform(remaining[['redshift_final','sum_p','u-r_p','g-i_p','r-z_p']]))
print(accuracy_score(rem_label, rem_pred))
print(classification_report(rem_label, rem_pred))

0.9995006776517583
              precision    recall  f1-score   support

      galaxy       1.00      1.00      1.00    481099
      quasar       1.00      1.00      1.00    139553
        star       1.00      1.00      1.00    318621

    accuracy                           1.00    939273
   macro avg       1.00      1.00      1.00    939273
weighted avg       1.00      1.00      1.00    939273



## Results from Inital Model
Evidently, it seems like the model really is predicting the each class perfectly, likely due to how the redshift works. We will proceed to utilize an SVM model to evaluate the performances of models against each other. Following this, we will move onto the image classification.

In [67]:
X_train, X_test, y_train, y_test = train_test_split(x_under, y_under, test_size=0.3, stratify=y_under, random_state=42)

id_train = X_train.drop('dr8objid',axis=1)
id_test = X_test.drop('dr8objid',axis=1)

X_train.drop('dr8objid',inplace=True,axis=1)
X_test.drop('dr8objid',inplace=True,axis=1)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [68]:
svm_model = SVC()
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [69]:
print(accuracy_score(y_test, y_pred_svm))
print(classification_report(y_test, y_pred_svm))

0.9948444444444444
              precision    recall  f1-score   support

      galaxy       1.00      0.99      0.99     15000
      quasar       1.00      1.00      1.00     15000
        star       0.99      1.00      0.99     15000

    accuracy                           0.99     45000
   macro avg       0.99      0.99      0.99     45000
weighted avg       0.99      0.99      0.99     45000



## Exploring results using PCA for preprocessing
Using the SVM we see the results are great, similar to our first model. We would now like to evaluate if we could maintain these results while reducing dimensionality via PCA. Our assumption is that the PCA will attribute its main component on the redshift since that has the greatest variance of any other variable and explains the biggest difference.

In [72]:
X_train, X_test, y_train, y_test = train_test_split(x_under, y_under, test_size=0.3, stratify=y_under, random_state=42)

id_train = X_train.drop('dr8objid',axis=1)
id_test = X_test.drop('dr8objid',axis=1)

X_train.drop('dr8objid',inplace=True,axis=1)
X_test.drop('dr8objid',inplace=True,axis=1)

scaler = StandardScaler()
pca = PCA(n_components=2)
pc_train = pca.fit_transform(scaler.fit_transform(X_train))
pc_test = pca.fit_transform(scaler.fit_transform(X_test))


svm_model2 = SVC()
svm_model2.fit(pc_train, y_train)
y_pred_svm2 = svm_model2.predict(pc_test)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [74]:
print(accuracy_score(y_test, y_pred_svm2))
print(classification_report(y_test, y_pred_svm2))

0.1834888888888889
              precision    recall  f1-score   support

      galaxy       0.00      0.00      0.00     15000
      quasar       0.00      0.00      0.00     15000
        star       0.68      0.55      0.61     15000

    accuracy                           0.18     45000
   macro avg       0.23      0.18      0.20     45000
weighted avg       0.23      0.18      0.20     45000



## PCA Results
Using PCA to reduce dimensionality, the svm model was retrained on using the resulting features from the principal components. Based on the results, it seems the PCA failed to produce the same results the previous models produced, so we will opt to utilize the RFC model for our initial classification. Now that we have evaluated multiple models, and tested some preprocessing techniques we can move on to the image classification using the prediction classifications from our initial model

In [78]:
remaining['pred_class'] = rem_pred

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_1993/1361899693.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  remaining['pred_class'] = rem_pred


In [81]:
remaining['pred_class']

0          galaxy
1          galaxy
3          galaxy
4          galaxy
5          galaxy
            ...  
1089266      star
1089267      star
1089269      star
1089270    quasar
1089271    quasar
Name: pred_class, Length: 939273, dtype: object